ultimatepython/classes/encapsulation.py

```diff
@@ -0,0 +1,108 @@
+"""
+Encapsulation is a feature of OOP that allows you to hide the
+implementation details of a class from its users.
+Encapsulation allows us to limit the access of certain attributes
+within a class. This prevents users from directly accessing and modifying such
+attributes from outside the class. Instead, users must use methods to access and
+modify attributes.
+"""
+import secrets
+
+
+class BankAccount:
+    def __init__(self, account_holder_name):
+        """
+        In python, a class attribute can be made private by prefixing it with two underscores.
+        This makes it inaccessible to users outside the class.
+        By default, class attributes are public. Therefore, they can be accessed and modified
+        outside of the class.
+
+        Here, account_number and balace are private while account_holder_name is public.
+        """
+        self.account_holder_name = account_holder_name
+        """
+        The account number is generated automaticallyusing the randint function from
+        the random module when a new instance of the class is created.
+        The balance is set to 0.0 by default.
+        """
+        self.__account_number = secrets.randbelow(10**10)  # generate a random account number of 10 digits.
+        self.__balance = 0.0
+
+    def deposit(self, balance):
+        self.__balance += balance  # add the deposited amount to the balance.
+
+    def withdraw(self, balance):
+        self.__balance -= balance
+
+    def get_balance(self):
+        return self.__balance
+
+    def get_account_number(self):
+        """
+        The account number is generated randomly when a new instance of the class is created.
+        Since the attribute is also private, it cannot be accessed directly from outside the class.
+        The get_account_number method allows you to access the account number outside of the class.
+        But since we do not define a setter method for this variable, we cannot modify it outside the class.
+        Therefore, the account number generated while creating an object of the BankAccount class cannot be changed
+        but can only be read using this function.
+        """
+        return self.__account_number
+
+    def __set_account_number(self, number):
+        """
+        This is a private method. Similar to private variables,
+        private methods also cannot be accessed outside the class.
+        """
+        self.__account_number = number
+
+    def delete_account(self):
+        """
+        This method is used to delete an account.
+        """
+        self.__balance = 0.0
+        self.__set_account_number(0)
+        self.account_holder_name = ""
+
+
+def main():
+    # Account names constants.
+    USER1 = "John Doe"
+    USER2 = "Jane Doe"
+
+    # Account instances.
+    account1 = BankAccount(USER1)
+    account2 = BankAccount(USER2)
+
+    # Accounts list.
+    accounts = [account1, account2]
+
+    assert account1.account_holder_name == USER1
```

Comment:
Didn't we mention in the lesson that we don't want to access attributes directly? Does it make sense to do it here just because it's "public"? Then people can freely modify the account name with no consequences

Reply:
I set one variable as to public to showcase the use case as well. We can do whatever, I will set it to private if you think that's correct. 😄

Reply:
Let's keep it as is then